components/backup-stream/src/router.rs

```diff
@@ -964,69 +988,167 @@ impl StreamTaskInfo {
-        let filepath = &data_file.storage_path;
-        let est_len = stat.len();
 
-        let ret = storage.write(filepath, reader, est_len).await;
+        let files_reader = FilesReader::new(data_files_open);
+
+        let reader = UnpinReader(Box::new(limiter.limit(files_reader.compat())));
+        let filepath = &merged_file_info.path;
+
+        let ret = storage.write(filepath, reader, stat_length).await;
+
         match ret {
             Ok(_) => {
                 debug!(
                     "backup stream flush success";
-                    "tmp file" => ?data_file.local_path,
                     "storage file" => ?filepath,
+                    "est_len" => ?stat_length,
                 );
             }
             Err(e) => {
                 warn!("backup stream flush failed";
-                    "file" => ?data_file.local_path,
-                    "est_len" => ?est_len,
+                    "est_len" => ?stat_length,
                     "err" => ?e,
                 );
                 return Err(Error::Io(e));
             }
         }
+
+        // push merged file into metadata
+        metadata.push(merged_file_info);
         Ok(())
     }
 
-    pub async fn flush_log(&self) -> Result<()> {
-        // if failed to write storage, we should retry write flushing_files.
+    pub async fn flush_log(&self, metadata: &mut MetadataInfoV2) -> Result<()> {
         let storage = self.storage.clone();
-        let files = self.flushing_files.write().await;
-
-        for batch_files in files.chunks(FLUSH_LOG_CONCURRENT_BATCH_COUNT) {
-            let futs = batch_files
-                .iter()
-                .map(|(_, v, _)| Self::flush_log_file_to(storage.clone(), v));
-            futures::future::try_join_all(futs).await?;
-        }
+        self.merge_log(metadata, storage.clone(), &self.flushing_files, false)
+            .await?;
+        self.merge_log(metadata, storage.clone(), &self.flushing_meta_files, true)
+            .await?;
 
         Ok(())
     }
 
-    pub async fn flush_meta(&self, metadata_info: MetadataInfo) -> Result<()> {
-        let meta_path = metadata_info.path_to_meta();
-        let meta_buff = metadata_info.marshal_to()?;
-        let buflen = meta_buff.len();
+    async fn merge_log(
+        &self,
+        metadata: &mut MetadataInfoV2,
+        storage: Arc<dyn ExternalStorage>,
+        files_lock: &RwLock<Vec<(TempFileKey, Mutex<DataFile>, DataFileInfo)>>,
+        is_meta: bool,
+    ) -> Result<()> {
+        let files = files_lock.write().await;
+        let mut batch_size = 0;
+        // file[batch_begin_index, i) is a batch
+        let mut batch_begin_index = 0;
+        // TODO: upload the merged file concurrently,
+        // then collect merged_file_infos and push them into `metadata`.
+        for (i, (_, _, info)) in files.iter().enumerate() {
+            if batch_size >= MERGED_FILE_SIZE_LIMIT {
```

Comment:
Can we replace the `MERGED_FILE_SIZE_LIMIT` for BackupStreamConfig.file_size_limit with the default value of 128M ?

Reply:
`file_size_limit`'s default value is 256MB, and `merged_file_size_limit`'s should larger than it. So set 256MB as `merged_file_size_limit`'s default value.